pkg/scheduler/queue/tenant_queues.go

```diff
@@ -190,17 +193,39 @@ func (qb *queueBroker) dequeueRequestForQuerier(
 // but there is no reason to make consumers know that they need to call through to the tenantQuerierAssignments.
 
 func (qb *queueBroker) addQuerierWorkerConn(conn *QuerierWorkerConn) (resharded bool) {
-	return qb.tenantQuerierAssignments.addQuerierWorkerConn(conn)
+	// if conn is for a new querier, we need to recompute tenant querier relationship; otherwise, we don't reshard
+	if addQuerier := qb.querierConnections.addQuerierWorkerConn(conn); addQuerier {
+		qb.tenantQuerierAssignments.addQuerier(conn.QuerierID)
+		return qb.tenantQuerierAssignments.recomputeTenantQueriers()
+	}
+	return false
+	//return qb.tenantQuerierAssignments.addQuerierWorkerConn(conn)
 }
 
 func (qb *queueBroker) removeQuerierWorkerConn(conn *QuerierWorkerConn, now time.Time) (resharded bool) {
-	return qb.tenantQuerierAssignments.removeQuerierWorkerConn(conn, now)
+	// if we're removing the last active connection for the querier, the querier may need to be removed
+	if removedQuerier := qb.querierConnections.removeQuerierWorkerConn(conn, now); removedQuerier {
+		return qb.tenantQuerierAssignments.removeQueriers(conn.QuerierID)
+	}
+	return false
 }
 
+// notifyQuerierShutdown handles a graceful shutdown notification from a querier.
+// Returns true if tenant-querier reshard was triggered.
 func (qb *queueBroker) notifyQuerierShutdown(querierID QuerierID) (resharded bool) {
-	return qb.tenantQuerierAssignments.notifyQuerierShutdown(querierID)
+	if removedQuerier := qb.querierConnections.shutdownQuerier(querierID); removedQuerier {
```

Comment:
same note as below - let `shutdownQuerier` do the removing. There's no case at the moment where the caller should do anything else, so don't give them the chance.

Reply:
also if we have these two usages of `removeQuerier`, moved behind the respective `shutdownQuerier` and `removeForgettableQueriers` interfaces, there are no `removeQuerier` usages outside the `querier_worker_conn_manager`.

If we are not going to draw a package boundary and use public/private methods to prevent unwanted usage, we could still just delete `removeQuerier` to show that it's not something that should be done outside the context of the pre-prepared use cases of shutting down or forgetting queriers.